pkg/planner/core/exhaust_physical_plans.go

```diff
@@ -1378,10 +1402,105 @@ func (p *LogicalJoin) constructInnerIndexScanTask(
+		return t
+	}
+	la, canPushAggToCop := wrapper.zippedChildren[len(wrapper.zippedChildren)-1].(*LogicalAggregation)
+	if la != nil && la.HasDistinct() {
+		// TODO: remove AllowDistinctAggPushDown after the cost estimation of distinct pushdown is implemented.
+		// If AllowDistinctAggPushDown is set to true, we should not consider RootTask.
+		if !la.SCtx().GetSessionVars().AllowDistinctAggPushDown {
+			canPushAggToCop = false
+		}
+	}
+
+	if canPushAggToCop {
+		// Try stream agg
+		_, preferStream := la.ResetHintIfConflicted()
+		for _, aggFunc := range la.AggFuncs {
+			if aggFunc.Mode == aggregation.FinalMode {
+				preferStream = false
+				break
+			}
+		}
+		// group by a + b is not interested in any order.
+		groupByCols := la.GetGroupByCols()
+		if len(groupByCols) != len(la.GroupByItems) {
+			preferStream = false
+		}
+		if la.HasDistinct() && !la.distinctArgsMeetsProperty() {
+			preferStream = false
+		}
+		// sort items must be the super set of group by items
+		if path != nil && path.Index != nil && ds.tableInfo.GetPartitionInfo() == nil {
+			if len(path.IdxCols) < len(groupByCols) {
+				preferStream = false
+			}
+			sctx := p.SCtx()
+			for i, groupbyCol := range groupByCols {
+				if path.IdxColLens[i] != types.UnspecifiedLength ||
+					!groupbyCol.EqualByExprAndID(sctx.GetExprCtx().GetEvalCtx(), path.IdxCols[i]) {
+					preferStream = false
+				}
+			}
+		} else {
+			preferStream = false
+		}
+
+		// build physical agg and attach to task
+		var aggTask Task
+		// build stream agg and change ds keep order to true
+		if preferStream {
+			newGbyItems := make([]expression.Expression, len(la.GroupByItems))
+			copy(newGbyItems, la.GroupByItems)
+			newAggFuncs := make([]*aggregation.AggFuncDesc, len(la.AggFuncs))
+			copy(newAggFuncs, la.AggFuncs)
+			streamAgg := basePhysicalAgg{
+				GroupByItems: newGbyItems,
+				AggFuncs:     newAggFuncs,
+			}.initForStream(la.SCtx(), la.StatsInfo(), la.QueryBlockOffset(), nil)
+			streamAgg.SetSchema(la.schema.Clone())
+			// change to keep order
+			physicalIndexScan, _ := dsCopTask.tablePlan.(*PhysicalIndexScan)
+			if physicalIndexScan != nil {
+				physicalIndexScan.KeepOrder = true
+				dsCopTask.keepOrder = true
+				aggTask = streamAgg.Attach2Task(dsCopTask)
+			}
+		}
+
+		// build hash agg, when the stream agg is illegal such as the order by prop is not matched
+		if aggTask == nil {
+			physicalHashAgg := NewPhysicalHashAgg(la, la.StatsInfo(), nil)
+			physicalHashAgg.SetSchema(la.schema.Clone())
+			aggTask = physicalHashAgg.Attach2Task(dsCopTask)
+		}
+
+		// build other inner plan node to task
+		result, ok := aggTask.(*RootTask)
+		if !ok {
+			return nil
+		}
+		result.p = p.constructInnerByZippedChildren(wrapper.zippedChildren[0:len(wrapper.zippedChildren)-1], result.p)
+		return result
+	}
+
+	result := dsCopTask.ConvertToRootTask(ds.SCtx())
+	result.SetPlan(p.constructInnerByZippedChildren(wrapper.zippedChildren, result.GetPlan()))
+	return result
```

Comment:
Looks like this case can be combined into the `if len(wrapper.zippedChildren) == 0` case above, just like the previous code, and can make the code structure slightly simpler.